### PR TITLE
Pandas 2

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---
Update dependency to use pandas 2.0. There also need to be some changes to pydantic implementations to avoid deprecation warnings and get it ready for 3.0.

Task Checklist
----
- [x] Have unittests been added (testing individual functions and their edge cases)
- [x] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [x] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

**Comment by isaac-tfn** [comment]
@MattBuckley-TfN Mainly getting pydantic 2 and pandas 2 in, as well as a couple of bug fixes, no rush really AECOM can work from the branch

File: src/caf/core/data_structures.py [hunk 1]
```diff
@@ -460,6 +479,16 @@ def translate_zoning(
             The weighting to use when building the translation. Must be
             one of TranslationWeighting.
 
+        check_totals: bool = True
+            Whether to raise a warning if the translated total doesn't match the
+            input total. Should be set to False for one-to-one translations.
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:484]: I think this is badly phrased, what I meant was it should be set to zero where you're using a translation of all 1s to disaggregate factors for example, so that the output would be larger than the input by design. That's a short term fix though, as I think a.) check_totals shouldn't raise an error, and b.) that option for disaggregating factors should be an explicit option rather than a workaround.

File: src/caf/core/data_structures.py [hunk 2]
```diff
@@ -460,6 +479,16 @@ def translate_zoning(
...
+
+        one-to-one: bool = False
+            Whether to run as a one-to-one translation, e.g. all data will be
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:487]: Similar to above comment - this is badly named and described, it means either many to one or disaggregating factors.

File: src/caf/core/data_structures.py [hunk 3]
```diff
@@ -635,6 +670,12 @@ def aggregate(self, segs: list[str]):
         segs: Segments to aggregate to. Must be a subset of self.segmentation.naming_order,
         naming order will be preserved.
         """
+        if not isinstance(segs, list):
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:673]: Good points, this section is actually changed in another PR so I'm not going to address here, but yes that is handled lower down and in the newer change it also allows a new Segmentation instance to be passed, rather than a list of string segment names.

File: src/caf/core/segmentation.py [hunk 4]
```diff
@@ -59,23 +59,25 @@ class SegmentationInput(BaseConfig):
     """
 
     enum_segments: list[SegmentsSuper]
-    naming_order: list[str]
-    subsets: dict[str, list[int]] = pydantic.Field(default_factory=dict)
     custom_segments: list[Segment] = pydantic.Field(default_factory=list)
+    subsets: dict[str, list[int]] = pydantic.Field(default_factory=dict)
+    naming_order: list[str]
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/segmentation.py:64]: I don't think pydantic cares about that, it's never complained about it for me before. I think I put these in a certain order for validators to work at some point but can be changed to make more sense now!

File: src/caf/core/segmentation.py [hunk 5]
```diff
@@ -59,23 +59,25 @@ class SegmentationInput(BaseConfig):
...
 
-    @pydantic.validator("subsets", always=True)
+    @pydantic.field_validator("subsets", check_fields=True)
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/segmentation.py:66]: Yep as mentioned this is done in ''functionality_for_tem" branch. If you're in a pydantic frame of mind you can have a look there, but it will go to you for review when done anyway

File: src/caf/core/zoning.py [hunk 6]
```diff
@@ -41,6 +41,8 @@ class TranslationWeighting(enum.Enum):
     EMPLOYMENT = "employment"
     NO_WEIGHT = "no_weight"
     AVERAGE = "average"
+    POP = "pop"
+    EMP = "emp"
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/zoning.py:45]: Ahh yes this was a hot fix because 'population' and 'employment' never actually appear in translation names as weightings so it wasn't finding them. I think I meant to chat to you about this as I think you added this bit initially but I never got round to it.

**Review by MattBuckley-TfN — APPROVED** [review]
Happy to merge this after our chat @isaac-tfn

All previous comments have been dealt with, are handled in another PR or have been logged as an issue
